AI Inheritance Refactor

fnnbrr opened the pull request:
AI/NPC classes will now be created as descendants of the abstract BaseAi class.  They will use Component classes (e.g. Patroller, Searcher, Chaser) to mix & match specific functionality, allowing for more modularity compared to a purely Inheritance-based structure.  This video is a great intro to how Component classes work: https://www.youtube.com/watch?v=8TIkManpEu4

Addresses Issues:
#214 
#215 
#212 
#228

Commit:
Moved most AI functions into BaseAi.cs

Commit:
First versions of NPC component classes

Commit:
First iteration of component-based AI system

Commit:
Fixed broken waiting function

Commit:
Fixed BasicNurse animation toggling and responder spawns

Commit:
Added unique movement speed to each state/component

Commit:
Added wait time support for each NPC waypoint

Commit:
Added custom rotations to each NPC waypoint

Commit:
Fixed Euler angles --> Quaternion conversion error

masonachu commented on Assets/Scripts/NPCs/BasicNurse.cs:
```diff
@@ -0,0 +1,122 @@
+                default:
+                    questionMark.SetActive(false);
+                    throw new UnityException("Invalid state name passed to " + GetType().Name);
+            }
```
Hey Finn!

I want to add a few FMOD params (float) for when you get caught so that the music dynamically changes when the nurse is searching and chasing. Could we chat about how we can put that in during your refactor?

fnnbrr replied:
Yeah no problem, once this PR is merged into master I can help you on a different branch to add FMOD stuff into the refactored AI system!

masonachu commented on Assets/Scripts/NPCs/BasicNurse.cs:
```diff
@@ -0,0 +1,122 @@
+
+        protected override void Update()
+        {
+            animator.SetBool(Constants.ANIMATION_STEVE_MOVING, !agent.isStopped);
```
I should mention that we have a walking animation as well sitting in the prefabs - maybe walking when patrolling/searching and running when chasing?

fnnbrr replied:
Sounds good, do you know what the name for that animation is in Constants? (e.g. Constants.ANIMATION_STEVE_WALKING ?)

jurcaua replied:
Walking isn't set up yet, so there aren't any animation states or logic or constants set up.
We can add walking in a separate PR, I can look into it after this is merged in since I did the running part of it.

Commit:
Created better patrol routes for kitchen AI

Commit:
Added new toggle animation system to stop/start NPCs

Commit:
Tried to avert some NavMeshAgent collision errors

FinnbarrOC commented:
NavMeshAgents still run into eachother and disrupt their patrols, so adding in some less braindead avoidance mechanisms can be in another PR. This one is mainly just a component-based refactor of what we already have, plus patrol waypoints with wait times and rotations.

Commit:
Merge branch 'master' into finnbarr

jurcaua commented on Assets/Prefabs/Props/Pine_Tree.prefab:
```diff
@@ -68,7 +68,7 @@ MeshRenderer:
   m_RendererPriority: 0
   m_Materials:
-  - {fileID: -3086382296381693183}
+  - {fileID: 2100000, guid: 4669023dba9b36242a67e8bac8b48e71, type: 2}
```
Just saw from checking out your branch myself but, can we make the Pine_Tree use an unlit shader instead? Probably don't need it to be included in lighting calculations.

jurcaua commented on Assets/Scripts/NPCs/BaseAi.cs:
```diff
@@ -0,0 +1,30 @@
+        public void Awake()
+        {
+            agent = Utils.GetRequiredComponent<NavMeshAgent>(this);
+            agent.avoidancePriority = Random.Range(0, 99);
```
Just curious personally, what's the effect of it being random?

fnnbrr replied:
Basically some of the spawned AI will have randomly higher priority, so those with lower priority will avoid them (leads to better pathfinding around each other).

jurcaua replied:
If the goal is to make sure everyone has a unique number (as much as possible), maybe we can have a static int instead where we do `agent.avoidancePriority = staticint++ % 99;`? Would be fancy and more deterministic (rather than having some play sessions sometimes having AIs with the same prio.

fnnbrr replied:
Great solution, just added it in!

jurcaua commented on Assets/Scripts/NPCs/BasicNurse.cs:
```diff
@@ -0,0 +1,126 @@
+    [RequireComponent(typeof(Animator))]
+    public class BasicNurse : BaseAi
+    {
+        private static readonly HashSet<string> States = new HashSet<string>()
```
Personally not a fan of using strings for the states... especially with all the string comparisons below. Seems like a red flag to me.
Is there a reason why you didn't go with just an enum?

fnnbrr replied:
Enums don't work with inheritance; the alternative is to declare one master enum of possible states in BaseAi and keep adding to it whenever we make new NPC types, but I thought that defeated the purpose of having inheritance in the first place.

jurcaua replied:
One thing maybe you can check out is the way I did the UI system with inheritance - basically where we can define a unique enum outside this class, and then make the BaseAI class templated on this enum. Check out `PauseMenuManager` for example! I think that may be cleaner personally.

fnnbrr replied:
Working on this now - I'm changing up all the NPC components so that instead of directly modifying the BasicNurse's behaviors (e..g changing state or destination), they just trigger events to the BasicNurse when important things happen

jurcaua replied:
That's awesome, I really like that. Looking forward to see that design get built up!!
Thanks for taking the time to refactor this all man

jurcaua commented on Assets/Scripts/NPCs/BasicNurse.cs:
```diff
@@ -0,0 +1,126 @@
+    [RequireComponent(typeof(Searcher))]
+    [RequireComponent(typeof(Patroller))]
+    [RequireComponent(typeof(Waiter))]
+    [RequireComponent(typeof(Animator))]
```
Seems like the animator could be on the child (from the `GetRequiredComponentInChildren`), so we would not need the component on the object this script is attached to (and this RequireComponent will auto-create it, which may not be needed then).

jurcaua commented on Assets/Scripts/NPCs/BasicNurse.cs:
```diff
@@ -0,0 +1,126 @@
+            waiter = Utils.GetRequiredComponent<Waiter>(this);
+            animator = Utils.GetRequiredComponentInChildren<Animator>(this);
+
+            questionMark = transform.Find("QuestionMark").gameObject;
```
Can we make the `questionMark` a public reference instead? Usually best practice not to use any of the Find w/ a string functions, since names commonly change.

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -0,0 +1,82 @@
+        
+        private List<Vector3> patrolPositions = new List<Vector3>();
+        private List<Quaternion> patrolRotations = new List<Quaternion>();
+        private List<float> patrolStayTimes = new List<float>();
```
I *think* (not sure) convention is to initialize these in Awake? I get yelled at by Visual Studio for this all the time so I have that ingrained in my mind now.

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            patrolPositions.Clear();
+            patrolRotations.Clear();
+            patrolStayTimes.Clear();
```
If you're just setting them to new values below anyways, you probably don't need these clears I don't think

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -0,0 +1,82 @@
+            if (patrolPositions.Count != patrolRotations.Count  || patrolPositions.Count != patrolStayTimes.Count)
+            {
+                throw new UnityException("Patroller passed waypoint information of differing lengths");
+            }
```
I think this kind of points to needing a List of some struct/class instead of keeping track of 3 separate lists and their lengths. Might be a cleaner design.

fnnbrr replied:
Good call, that cleaned things up a lot!

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -0,0 +1,82 @@
+            curStayTime = patrolStayTimes[0];
+        }
+
+        public bool RotationComplete()
```
Name seems kind of misleading here, like it would only return if the rotation is complete or not, but it actually does the rotation and deals with animations too! Seeing how it's used in the Update loop, maybe this additional logic can be according to the return value of this?

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -0,0 +1,82 @@
+
+            int newIndex;
+            if (loop) newIndex = destinationCount % patrolPositions.Count;
+            else newIndex = Utils.PingPong(destinationCount, patrolPositions.Count - 1);
```
I'd either go with expanding this to using curly brackets, or using a ternary operator instead. Kind of hard to read now for me personally.

jurcaua commented on Assets/Scripts/NPCs/Components/Responder.cs:
```diff
@@ -0,0 +1,55 @@
+    {
+        public float movementSpeed = 8.0f;
+        public Vector3 responsePoint;
+        public bool initiallyResponding = false;
```
Do we still need to support this `initiallyResponding` functionality? 
Also seems like something the component shouldn't care about, and be handled in the Nurse logic instead (like the guy said in the video you sent, this branching logic is something we want to avoid with a composition-based approach).

jurcaua commented on Assets/Scripts/NPCs/Components/Responder.cs:
```diff
@@ -0,0 +1,55 @@
+            if (relatedSearcher != null)
+            {
+                relatedSearcher.searchBoundsCenter = responsePoint;
+            }
```
Similar to the comment I left above, the checking if we're also a Searcher seems to be going against the composition pattern... 
Instead, possibly we can tell the BaseAI about needing to respond somewhere, and then the BaseAI can spread the word to other components, such as the Searcher in this case.

jurcaua commented on Assets/Scripts/NPCs/Components/Searcher.cs:
```diff
@@ -0,0 +1,38 @@
+            Vector3 randomPoint = searchBoundsCenter + Random.insideUnitSphere * searchBoundsRadius;
+            NavMeshHit navHit;
+
+            if (!NavMesh.SamplePosition(randomPoint, out navHit, searchBoundsRadius, NavMesh.AllAreas))
```
```suggestion
            if (!NavMesh.SamplePosition(randomPoint, out NavMeshHit navHit, searchBoundsRadius, NavMesh.AllAreas))
```

jurcaua commented on Assets/Scripts/NPCs/Components/Waiter.cs:
```diff
@@ -0,0 +1,38 @@
+            baseAi.ToggleAnimations(true);
+            waitTimer = 0f;
+            return true;
+        }
```
Could we maybe try something like, instead of continuously adding to a variable, set the variable to a future point of time and wait until Time.time equals that value? (ex, `float timerEndTime = Time.time + waitDuration;`)
The issue I see with this function (and the issue I had before when I was trying to work with it previously) was that this timer only counted when we called the function, so if something like, messes up for some reason or we need to disable the NavMeshAgent or something where we go down this path in the Update:
`if (agent.hasPath || !agent.isOnNavMesh || agent.pathPending) return;`

Then the wait isn't going and adding anymore.

Something like a coroutine to count for you independently and then sets a flag or calls a function or something is even better personally.

jurcaua replied:
Coroutine would be nice since then we don't even need to enter the update! And we can freely disable then enable the navmeshagent and all

fnnbrr replied:
Yeah that sounds ideal, I've seen you use the event/delegate system in other places so I'll do some reading and try to use that here (and for Responder in response to your comment there)

jurcaua replied:
That would be amazing! Thanks man 😄

jurcaua commented on Assets/Scripts/NPCs/PureChaser.cs:
```diff
@@ -1,105 +1,111 @@
-    private void ChaseTarget()
-    {
-        if (shouldChase && targetTransform)
+        private new void Awake()
```
Wait why do we need this to be a `new` function? Can we just make it override (and make the Awake function in BaseAI virtual)? Same for the Nurse script as well. From what I know, `new` is like hiding the base implementation all together and not commonly used usually... but I don't really know too much about this topic

jurcaua reviewed (changes requested):
Hey!! Tested it out locally and it's all working buttery smooth 🦋 
Thanks for all the hard work on the refactor, 🎆  loving it
got some comments for ya, mostly design related!!

sidenote, can we make it possible to look over the middle table somehow? not sure about the best way to handle that is but maybe just unmarking them as obstacle for now (visually maybe itll work odd but.... probably fine?)

Commit:
First round of fixes for PR requested changes

Commit:
Modified Patrollers to use lists of PatrolWaypoints

Replaces the old system of having 3 separate lists of positions/rotations/staytimes

Commit:
Decoupled Responder and Searcher NPC components

Commit:
Made BaseAi agent avoidance priority deterministic on spawn order

Commit:
Modified NPC classes to use generics and Enums for states

Commit:
Modified Chaser NPC component to communicate with Events

Commit:
Modified Responder and Searcher to use Events

Commit:
Modified NPC system to be Event-based

Commit:
Added reacquireInterval to Chaser to adjust aggression of chasing

Lower interval means the Chaser updates to follow the player's exact position more frequently, making it harder to outrun.

Commit:
Merge branch 'master' into finnbarr

FinnbarrOC commented:
Alrighty @jurcaua, take a look & see how it runs for ya!
This is my first time using Events/Delegates, so any advice on that end would be great.

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -18,38 +19,34 @@ public class Patroller : MonoBehaviour
-            // agent.SetDestination(patrolWaypoints[0].position); TODO replace with Event
-            curRotation = Quaternion.Euler(patrolWaypoints[0].rotation);
-            curStayTime = patrolWaypoints[0].stayTime;
+            curPatrolWaypoint = patrolWaypoints[0];
```
Check the size of the list before doing this?

jurcaua commented on Assets/Scripts/NPCs/Components/Patroller.cs:
```diff
@@ -18,38 +19,34 @@ public class Patroller : MonoBehaviour
-                return true;
+                transform.rotation = Quaternion.RotateTowards(transform.rotation, curRotation, 
+                    Time.deltaTime * rotationSpeed);
+                yield return new WaitForSeconds(Time.deltaTime);
```
Are the rotations choppy (I haven't checked yet)? If so, it may be because we're using deltaTime here. Switching to fixedDeltaTime SHOULD fix any lag issues here

fnnbrr replied:
I couldn't see any visual difference between deltaTime and fixedDeltaTime, but I swapped both to fixedDeltaTime just for consistency's sake if our framerates ever tank.

Commit:
Update MAIN.unity

jurcaua reviewed (approved):
Everything is looking amazing now wow, tested locally for me and it all works as expected!
Code looking top notch as well, loving it 😎 
v small comments below + discord messages but otherwise, should be good to go!

Commit:
Final tweaks to Patroller for consistency
